### PR TITLE
feat: Anvil dnd enhancement of highlight info and highlight skipping.

### PR DESCRIPTION
**Pull request by marks0351**
> Pull Request Template
>
> Use this template to quickly create a well written pull request. Delete all quotes before creating the pull request.
>
## Description
In this PR we are enhancing/changing a couple of things
- change skipping highlight logic.
  - we skip showing vertical highlights when the only hug widget in a row is dragged, now we show vertical highlights.
  - instead of skipping the vertical highlight left to a widget which is being dragged, we skip the highlight right to the widget so that changing order of widgets is more intutive.
- adding new prop to highlight info to mark highlights of the existing dragged widget so that such drops are not considered to update widgets/layouts. This will make sure undo / redo works as expected.

#### PR fixes following issue(s)
Fixes # (issue number)
> if no issue exists, please create an issue and ask the maintainers about this first
>
>
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
	- Enabled `ab_wds_enabled` and `release_anvil_enabled` feature flags for all users.
	- Enhanced the canvas interaction in the Anvil layout system by introducing logic to better handle horizontal and vertical highlights, improving the user experience during widget drag-and-drop operations.
	- Modified the logic for determining overlay widgets, streamlining the process for identifying modal and similar widgets.

- **Refactor**
	- Updated feature flag selection logic to return default values more accurately.
	- Refined highlight generation logic in Anvil layout system to account for dragged widgets and discarded layouts, offering a more intuitive drag-and-drop experience.
	- Adjusted the `getLayoutSystemType` selector to consistently return `LayoutSystemTypes.ANVIL`, ensuring a unified layout system experience.

- **Tests**
	- Enhanced test coverage for highlight selection algorithms in the Anvil layout system, ensuring accurate highlight behavior based on mouse position.

### CODE REVIEW

File: app/client/src/ce/entities/FeatureFlag.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ export const DEFAULT_FEATURE_FLAG_VALUE: FeatureFlags = {
   release_datasource_environments_enabled: false,
   release_appnavigationlogoupload_enabled: false,
   release_embed_hide_share_settings_enabled: false,
-  ab_wds_enabled: false,
```
Comment (marks0351) [inline app/client/src/ce/entities/FeatureFlag.ts:65]: changes to enable anvil temporarily will be removed when the PR is done testing.

File: app/client/src/ce/selectors/featureFlagsSelectors.ts [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
-import type { FeatureFlag } from "@appsmith/entities/FeatureFlag";
+import {
+  DEFAULT_FEATURE_FLAG_VALUE,
+  type FeatureFlag,
+} from "@appsmith/entities/FeatureFlag";
 import { createSelector } from "reselect";
 import { FEATURE_FLAG } from "@appsmith/entities/FeatureFlag";
 
 export const selectFeatureFlags = (state: AppState) =>
-  state.ui.users.featureFlag.data;
```
Comment (marks0351) [inline app/client/src/ce/selectors/featureFlagsSelectors.ts:7]: changes to enable anvil temporarily will be removed when the PR is done testing.

File: app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx [hunk 3]
```diff
@@ -42,30 +44,20 @@ export const AnvilFlexComponent = forwardRef(
...
         | (Partial<WidgetProps> & WidgetConfigProps & { type: string })
         | undefined = WidgetFactory.getConfig(widgetType);
-      if (!widgetConfig) return;
-      setIsFillWidget(
-        widgetConfig?.responsiveBehavior === ResponsiveBehavior.Fill,
-      );
-      setVerticalAlignment(
-        widgetConfig?.flexVerticalAlignment || FlexVerticalAlignment.Top,
-      );
+      const isFillWidget =
```
Comment (marks0351) [inline app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx:51]: @prsidhu I have removed this state, let me know if I failed to take into consideration a particular scenario.

File: app/client/src/reducers/uiReducers/dragResizeReducer.ts [hunk 4]
```diff
@@ -102,7 +102,9 @@ export const widgetDraggingReducer = createImmerReducer(initialState, {
     state: WidgetDragResizeState,
     action: ReduxAction<{ widgetId?: string }>,
   ) => {
-    state.focusedWidget = action.payload.widgetId;
+    if (state.focusedWidget !== action.payload.widgetId) {
```
Comment (marks0351) [inline app/client/src/reducers/uiReducers/dragResizeReducer.ts:105]: @hetunandu I have added this check coz everytime the main canvas was clicked, the widget pane kept rerendering

File: app/client/src/selectors/layoutSystemSelectors.ts [hunk 5]
```diff
@@ -1,18 +1,8 @@
...
-      ?.type
-  ) {
-    return LayoutSystemTypes[
-      state.ui.applications.currentApplication?.applicationDetail
-        ?.appPositioning?.type
-    ];
-  }
-  return LayoutSystemTypes.FIXED;
+export const getLayoutSystemType = () => {
+  return LayoutSystemTypes.ANVIL;
```
Comment (marks0351) [inline app/client/src/selectors/layoutSystemSelectors.ts:7]: changes to enable anvil temporarily will be removed when the PR is done testing.

File: app/client/src/reducers/uiReducers/dragResizeReducer.ts (hunk 4 above)
Comment (hetunandu) [reply app/client/src/reducers/uiReducers/dragResizeReducer.ts:105]: LGTM

**Comment by prsidhu** [comment]
@marks0351 Unable to move a hug widget within the same row, if it is the only widget in that row. 

https://github.com/appsmithorg/appsmith/assets/5424788/be156e84-4e9f-4e33-bf42-3248f99c87f6

**Review by prsidhu — APPROVED** [review]

**Review by prsidhu — APPROVED** [review]

**Pull request merged**